File: C10_Lights/LightSequencerCS/NewPlaylistForm.cs
//////////////////////////////////////////////////////////////////////////////////
//	NewPlaylistForm.cs
//	Light Sequencer
//	Written by Brian Peek (http://www.brianpeek.com/)
//	for the Animated Holiday Lights article
//		at Coding4Fun (http://msdn.microsoft.com/coding4fun/)
//////////////////////////////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;
using System.IO;
using System.Windows.Forms;

namespace LightSequencer
{
	public partial class NewPlaylistForm : Form
	{
		public struct ListBoxSequence
		{
			public Sequence Sequence;
			public string File;

			public ListBoxSequence(string file, Sequence sequence)
			{
				File = file;
				Sequence = sequence;
			}

			public override string ToString()
			{
				return string.IsNullOrEmpty(Sequence.Title) ? Path.GetFileName(File) : Sequence.Title;
			}
		}

		private SequencePlayer _player;
		private int _index;
		private bool playing;

		public NewPlaylistForm() : this(null)
		{
		}

		public NewPlaylistForm(string file)
		{
			InitializeComponent();

			if(!string.IsNullOrEmpty(file))
			{
				Playlist p = new Playlist(file);
				RefreshList(p.Filenames);
			}
		}

		private void RefreshList(IList<string> files)
		{
			foreach(string file in files)
			{
				Sequence s = new Sequence(file);
				lbSequences.Items.Add(new ListBoxSequence(file, s));
			}
		}

		private void btnAdd_Click(object sender, EventArgs e)
		{
			OpenFileDialog ofd = new OpenFileDialog();
			ofd.Multiselect = true;
			ofd.Title = "Locate sequences...";
			ofd.Filter = "Sequences (*.seq)|*.seq|All files (*.*)|*.*";
			if(ofd.ShowDialog() == DialogResult.OK)
				RefreshList(ofd.FileNames);
		}

		private void btnUp_Click(object sender, EventArgs e)
		{
			// Get the collection of selected indexes in the ListBox.  
			ListBox.SelectedIndexCollection sic = this.lbSequences.SelectedIndices;
		 
			// Loop through the ListBox.
			for (int i = 0; i <= this.lbSequences.Items.Count - 1; i++)
			{
				// If this index is in the collection.
				if ((sic.Contains(i)) && (i > 0))
				{
					// Move the item one index upward.
					object item = this.lbSequences.Items[i];
					this.lbSequences.Items.RemoveAt(i);
					this.lbSequences.Items.Insert(i-1, item);
					this.lbSequences.SetSelected(i-1, true);
				}
			}
		}

		private void btnSave_Click(object sender, EventArgs e)
		{
			DialogResult dr = DialogResult.OK;

			SaveFileDialog sfd = new SaveFileDialog();
			sfd.Title = "Save Playlist File...";
			sfd.Filter = "Light Sequencer Playlists (*.lsp)|*.lsp";
			sfd.AddExtension = true;
			dr = sfd.ShowDialog();

			if(dr == DialogResult.OK)
			{
				Playlist p = new Playlist();
				foreach(ListBoxSequence lbs in lbSequences.Items)
					p.Filenames.Add(lbs.File);
				p.Save(sfd.FileName);
				MessageBox.Show("Playlist saved.", "Saved", MessageBoxButtons.OK, MessageBoxIcon.Information);
			}
		}

		private void btnPlay_Click(object sender, EventArgs e)
		{
			if(!playing)
			{
				btnPlay.Text = "&Stop";
				_index = 0;
				PlaySong(_index);
			}
			else
			{
				btnPlay.Text = "&Play";
				lblStatus.Text = "Stopped...";
				_player.Unload();
			}

			gbPlaylist.Enabled = playing;
			playing = !playing;
			btnNext.Enabled = btnPrev.Enabled = playing;
		}

		void _sp_SequenceStopped(object sender, EventArgs e)
		{
			this._player.Unload();

			// move to the next song
			if(_index < lbSequences.Items.Count-1)
				_index++;
			else
			{
				// if repeat checked, start over, otherwse get out and don't play a song
				if(chkRepeat.Checked)
					_index = 0;
				else
				{
					// stop the list
					this.BeginInvoke(new MethodInvoker(delegate() {btnPlay.PerformClick();}));
					return;
				}
			}

			PlaySong(_index);
		}

		void PlaySong(int index)
		{
			ListBoxSequence lbs = (ListBoxSequence)lbSequences.Items[index];
			this.BeginInvoke(new MethodInvoker(delegate(){lblStatus.Text = "Playing " + lbs.ToString();}));

			switch(lbs.Sequence.Version)
			{
				case 1:
					this._player = new SequencePlayerV1(lbs.Sequence);
					this._player.SequenceStopped += new EventHandler(_sp_SequenceStopped);
					this._player.Start();
					break;
				case 2:
					this._player = new SequencePlayerV2(lbs.Sequence);
					this._player.SequenceStopped += new EventHandler(_sp_SequenceStopped);
					this._player.Start();
					break;
			}
		}

		private void btnNext_Click(object sender, EventArgs e)
		{
			this._player.Unload();
			if(_index < lbSequences.Items.Count-1)
				_index++;
			else
				_index = 0;

			PlaySong(_index);
		}

		private void btnPrev_Click(object sender, EventArgs e)
		{
			this._player.Unload();
			if(_index == 0)
				_index = lbSequences.Items.Count-1;
			else
				_index--;

			PlaySong(_index);
		}

		private void btnDown_Click(object sender, EventArgs e)
		{
			ListBox.SelectedIndexCollection col = this.lbSequences.SelectedIndices; 

			for (int i = this.lbSequences.Items.Count - 1; i >= 0; i--)
			{
				if ((col.Contains(i)) && (i < this.lbSequences.Items.Count-1))
				{
					object item = this.lbSequences.Items[i];
					this.lbSequences.Items.RemoveAt(i);
					this.lbSequences.Items.Insert(i+1, item);
					this.lbSequences.SetSelected(i+1, true);
				}
			}
		}

		private void btnRemove_Click(object sender, EventArgs e)
		{
			this.lbSequences.Items.RemoveAt(lbSequences.SelectedIndex);
		}
	}
}

File: C10_Lights/LightSequencerVB/NewPlaylistForm.vb
'////////////////////////////////////////////////////////////////////////////////
'	NewPlaylistForm.cs
'	Light Sequencer
'	Written by Brian Peek (http://www.brianpeek.com/)
'	for the Animated Holiday Lights article
'		at Coding4Fun (http://msdn.microsoft.com/coding4fun/)
'////////////////////////////////////////////////////////////////////////////////


Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Windows.Forms

Namespace LightSequencer
	Partial Public Class NewPlaylistForm
		Inherits Form
		Public Structure ListBoxSequence
			Public Sequence As Sequence
			Public File As String

			Public Sub New(ByVal file As String, ByVal sequence As Sequence)
				Me.File = file
				Sequence = sequence
			End Sub

			Public Overrides Function ToString() As String
				If String.IsNullOrEmpty(Sequence.Title) Then
					Return Path.GetFileName(File)
				Else
					Return Sequence.Title
				End If
			End Function
		End Structure

		Private _player As SequencePlayer
		Private _index As Integer
		Private playing As Boolean

		Public Sub New()
			Me.New(Nothing)
		End Sub

		Public Sub New(ByVal file As String)
			InitializeComponent()

			If (Not String.IsNullOrEmpty(file)) Then
				Dim p As New Playlist(file)
				RefreshList(p.Filenames)
			End If
		End Sub

		Private Sub RefreshList(ByVal files As IList(Of String))
			For Each file As String In files
				Dim s As New Sequence(file)
				lbSequences.Items.Add(New ListBoxSequence(file, s))
			Next file
		End Sub

		Private Sub btnAdd_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnAdd.Click
			Dim ofd As New OpenFileDialog()
			ofd.Multiselect = True
			ofd.Title = "Locate sequences..."
			ofd.Filter = "Sequences (*.seq)|*.seq|All files (*.*)|*.*"
			If ofd.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
				RefreshList(ofd.FileNames)
			End If
		End Sub

		Private Sub btnUp_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnUp.Click
			' Get the collection of selected indexes in the ListBox.  
			Dim sic As ListBox.SelectedIndexCollection = Me.lbSequences.SelectedIndices

			' Loop through the ListBox.
			For i As Integer = 0 To Me.lbSequences.Items.Count - 1
				' If this index is in the collection.
				If (sic.Contains(i)) AndAlso (i > 0) Then
					' Move the item one index upward.
					Dim item As Object = Me.lbSequences.Items(i)
					Me.lbSequences.Items.RemoveAt(i)
					Me.lbSequences.Items.Insert(i-1, item)
					Me.lbSequences.SetSelected(i-1, True)
				End If
			Next i
		End Sub

		Private Sub btnSave_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnSave.Click
			Dim dr As DialogResult = System.Windows.Forms.DialogResult.OK

			Dim sfd As New SaveFileDialog()
			sfd.Title = "Save Playlist File..."
			sfd.Filter = "Light Sequencer Playlists (*.lsp)|*.lsp"
			sfd.AddExtension = True
			dr = sfd.ShowDialog()

			If dr = System.Windows.Forms.DialogResult.OK Then
				Dim p As New Playlist()
				For Each lbs As ListBoxSequence In lbSequences.Items
					p.Filenames.Add(lbs.File)
				Next lbs
				p.Save(sfd.FileName)
				MessageBox.Show("Playlist saved.", "Saved", MessageBoxButtons.OK, MessageBoxIcon.Information)
			End If
		End Sub

		Private Sub btnPlay_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnPlay.Click
			If (Not playing) Then
				btnPlay.Text = "&Stop"
				_index = 0
				PlaySong(_index)
			Else
				btnPlay.Text = "&Play"
				lblStatus.Text = "Stopped..."
				_player.Unload()
			End If

			gbPlaylist.Enabled = playing
			playing = Not playing
			btnPrev.Enabled = playing
			btnNext.Enabled = btnPrev.Enabled
		End Sub

		Private Sub _sp_SequenceStopped(ByVal sender As Object, ByVal e As EventArgs)
			Me._player.Unload()

			' move to the next song
			If _index < lbSequences.Items.Count-1 Then
				_index += 1
			Else
				' if repeat checked, start over, otherwse get out and don't play a song
				If chkRepeat.Checked Then
					_index = 0
				Else
					' stop the list
					Me.BeginInvoke(New MethodInvoker(AddressOf AnonymousMethod1))
					Return
				End If
			End If

			PlaySong(_index)
		End Sub
		Private Sub AnonymousMethod1()
			btnPlay.PerformClick()
		End Sub

		Private Sub PlaySong(ByVal index As Integer)
			Dim lbs As ListBoxSequence = CType(lbSequences.Items(index), ListBoxSequence)
''TODO: INSTANT VB TODO TASK: Anonymous methods are not converted by Instant VB if local variables of the outer method are referenced within the anonymous method:
'			Me.BeginInvoke(New MethodInvoker(delegate()
'				lblStatus.Text = "Playing " & lbs.ToString()
'		   ))

			Select Case lbs.Sequence.Version
				Case 1
					Me._player = New SequencePlayerV1(lbs.Sequence)
					AddHandler _player.SequenceStopped, AddressOf _sp_SequenceStopped
					Me._player.Start()
				Case 2
					Me._player = New SequencePlayerV2(lbs.Sequence)
					AddHandler _player.SequenceStopped, AddressOf _sp_SequenceStopped
					Me._player.Start()
			End Select
		End Sub

		Private Sub btnNext_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnNext.Click
			Me._player.Unload()
			If _index < lbSequences.Items.Count-1 Then
				_index += 1
			Else
				_index = 0
			End If

			PlaySong(_index)
		End Sub

		Private Sub btnPrev_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnPrev.Click
			Me._player.Unload()
			If _index = 0 Then
				_index = lbSequences.Items.Count-1
			Else
				_index -= 1
			End If

			PlaySong(_index)
		End Sub

		Private Sub btnDown_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnDown.Click
			Dim col As ListBox.SelectedIndexCollection = Me.lbSequences.SelectedIndices

			For i As Integer = Me.lbSequences.Items.Count - 1 To 0 Step -1
				If (col.Contains(i)) AndAlso (i < Me.lbSequences.Items.Count-1) Then
					Dim item As Object = Me.lbSequences.Items(i)
					Me.lbSequences.Items.RemoveAt(i)
					Me.lbSequences.Items.Insert(i+1, item)
					Me.lbSequences.SetSelected(i+1, True)
				End If
			Next i
		End Sub

		Private Sub btnRemove_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnRemove.Click
			Me.lbSequences.Items.RemoveAt(lbSequences.SelectedIndex)
		End Sub
	End Class
End Namespace
